File: client/src/main/java/pro/panopticon/client/model/Measurement.java
package pro.panopticon.client.model;

import com.amazonaws.services.cloudwatch.model.StandardUnit;

public class Measurement {
    public String key;
    public String status;
    public String displayValue;
    public CloudwatchValue cloudwatchValue;
    public String description;

    public Measurement(String key, String status, String displayValue, CloudwatchValue cloudwatchValue, String description) {
        this.key = key;
        this.status = status;
        this.displayValue = displayValue;
        this.cloudwatchValue = cloudwatchValue;
        this.description = description;
    }

    public Measurement(String key, String status, String displayValue, CloudwatchValue cloudwatchValue) {
        this(key, status, displayValue, cloudwatchValue, null);
    }

    public Measurement(String key, String status, String displayValue) {
        this(key, status, displayValue, null, null);
    }

    @Override
    public String toString() {
        return "Measurement{" +
                "key='" + key + '\'' +
                ", status='" + status + '\'' +
                ", displayValue='" + displayValue + '\'' +
                ", cloudwatchValue=" + cloudwatchValue +
                ", description=" + description +
                '}';
    }

    public static class CloudwatchValue {
        public final double value;
        public final StandardUnit unit;

        public CloudwatchValue(double value, StandardUnit unit) {
            this.value = value;
            this.unit = unit;
        }

    }
}

File: client/src/main/java/pro/panopticon/client/model/Measurement.kt
package pro.panopticon.client.model

import com.amazonaws.services.cloudwatch.model.StandardUnit

data class Measurement(
    val key: String,
    val status: Status,
    val displayValue: String,
    val cloudwatchValue: CloudwatchValue? = null,
    val description: String,
) {
    constructor(
        key: String,
        status: Status,
        displayValue: String,
        description: String,
    ) : this(key, status, displayValue, null, description)

    @Deprecated("String based status is deprecated. Use Status enum instead.")
    constructor(
        key: String,
        status: String,
        displayValue: String,
        description: String,
    ) : this(key, Status.fromString(status), displayValue, null, description)

    @Deprecated("String based status is deprecated. Use Status enum instead.")
    constructor(
        key: String,
        status: String,
        displayValue: String,
        cloudwatchValue: CloudwatchValue?,
        description: String,
    ) : this(key, Status.fromString(status), displayValue, cloudwatchValue, description)

    data class CloudwatchValue @JvmOverloads constructor(
        val value: Double,
        val unit: StandardUnit,
        val dimensions: List<MetricDimension> = emptyList(),
    )

    enum class Status {
        INFO,
        WARN,
        ERROR;

        companion object {
            fun fromString(str: String): Status {
                return when (str) {
                    "INFO" -> INFO
                    "WARN" -> WARN
                    "ERROR" -> ERROR
                    else -> ERROR
                }
            }
        }
    }
}
